In [ ]:
!pip install tensorflow

In [ ]:
!pip install --upgrade keras

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 17.3 MB/s eta 0:00:00
  Attempting uninstall: keras
    Found existing installation: keras 3.8.0
    Uninstalling keras-3.8.0:
      Successfully uninstalled keras-3.8.0


In [ ]:
!pip install datasets

In [ ]:
!huggingface-cli login

⚠️  Warning: 'huggingface-cli login' is deprecated. Use 'hf auth login' instead.

    _|    _|  _|    _|    _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|_|_|_|    _|_|      _|_|_|  _|_|_|_|
    _|    _|  _|    _|  _|        _|          _|    _|_|    _|  _|            _|        _|    _|  _|        _|
    _|_|_|_|  _|    _|  _|  _|_|  _|  _|_|    _|    _|  _|  _|  _|  _|_|      _|_|_|    _|_|_|_|  _|        _|_|_|
    _|    _|  _|    _|  _|    _|  _|    _|    _|    _|    _|_|  _|    _|      _|        _|    _|  _|        _|
    _|    _|    _|_|      _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|        _|    _|    _|_|_|  _|_|_|_|

    To log in, `huggingface_hub` requires a token generated from https://huggingface.co/settings/tokens .
Enter your token (input will not be visible): 

In [ ]:
#Download BigEarthNet dataset
from datasets import load_dataset
ds = load_dataset("danielz01/BigEarthNet-S2-v1.0")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/1.04k [00:00<?, ?B/s]

s2-rgb/test-00000-of-00007-fa75b96ba0241(…):   0%|          | 0.00/463M [00:00<?, ?B/s]

s2-rgb/test-00001-of-00007-7746faa4c2987(…):   0%|          | 0.00/463M [00:00<?, ?B/s]

s2-rgb/test-00002-of-00007-3df6bac341e50(…):   0%|          | 0.00/496M [00:00<?, ?B/s]

s2-rgb/test-00003-of-00007-134378bdb2c97(…):   0%|          | 0.00/472M [00:00<?, ?B/s]

s2-rgb/test-00004-of-00007-ba56718a154a3(…):   0%|          | 0.00/477M [00:00<?, ?B/s]

s2-rgb/test-00005-of-00007-db24561286ec3(…):   0%|          | 0.00/497M [00:00<?, ?B/s]

s2-rgb/test-00006-of-00007-2d2952dd94da4(…):   0%|          | 0.00/488M [00:00<?, ?B/s]

s2-rgb/val-00000-of-00007-f0643798602cca(…):   0%|          | 0.00/455M [00:00<?, ?B/s]

s2-rgb/val-00001-of-00007-04b0d699da1eb9(…):   0%|          | 0.00/455M [00:00<?, ?B/s]

s2-rgb/val-00002-of-00007-fdf80718140deb(…):   0%|          | 0.00/486M [00:00<?, ?B/s]

s2-rgb/val-00003-of-00007-e34a5e008b9d45(…):   0%|          | 0.00/465M [00:00<?, ?B/s]

s2-rgb/val-00004-of-00007-d7d1605cae3e62(…):   0%|          | 0.00/470M [00:00<?, ?B/s]

s2-rgb/val-00005-of-00007-67da670d67109d(…):   0%|          | 0.00/488M [00:00<?, ?B/s]

s2-rgb/val-00006-of-00007-e8aa45ccad7d0d(…):   0%|          | 0.00/480M [00:00<?, ?B/s]

s2-rgb/train-00000-of-00015-d6560d607332(…):   0%|          | 0.00/463M [00:00<?, ?B/s]

s2-rgb/train-00001-of-00015-9a124129340e(…):   0%|          | 0.00/462M [00:00<?, ?B/s]

s2-rgb/train-00002-of-00015-288e311ae08e(…):   0%|          | 0.00/462M [00:00<?, ?B/s]

s2-rgb/train-00003-of-00015-1ccc32c48471(…):   0%|          | 0.00/462M [00:00<?, ?B/s]

s2-rgb/train-00004-of-00015-36361aea05da(…):   0%|          | 0.00/488M [00:00<?, ?B/s]

s2-rgb/train-00005-of-00015-90829c95f348(…):   0%|          | 0.00/501M [00:00<?, ?B/s]

s2-rgb/train-00006-of-00015-3c4a1537f29b(…):   0%|          | 0.00/480M [00:00<?, ?B/s]

s2-rgb/train-00007-of-00015-11e2fb6c06dd(…):   0%|          | 0.00/467M [00:00<?, ?B/s]

s2-rgb/train-00008-of-00015-34d29c632d99(…):   0%|          | 0.00/465M [00:00<?, ?B/s]

s2-rgb/train-00009-of-00015-d2d8d6d6d458(…):   0%|          | 0.00/473M [00:00<?, ?B/s]

s2-rgb/train-00010-of-00015-ca2886dc8cd5(…):   0%|          | 0.00/497M [00:00<?, ?B/s]

s2-rgb/train-00011-of-00015-9a1860b8270f(…):   0%|          | 0.00/497M [00:00<?, ?B/s]

s2-rgb/train-00012-of-00015-2eacca1c13d3(…):   0%|          | 0.00/492M [00:00<?, ?B/s]

s2-rgb/train-00013-of-00015-d522c508701a(…):   0%|          | 0.00/481M [00:00<?, ?B/s]

s2-rgb/train-00014-of-00015-df0029906d78(…):   0%|          | 0.00/494M [00:00<?, ?B/s]

Generating test split:   0%|          | 0/125866 [00:00<?, ? examples/s]

Generating val split:   0%|          | 0/123723 [00:00<?, ? examples/s]

Generating train split:   0%|          | 0/269695 [00:00<?, ? examples/s]

In [ ]:
#This method converts the image labels from the 43-class version to the 19-class version
def map_labels(sample):
    label = sample["labels"]
    for i in range(len(label)):
      if label[i] == "Continuous urban fabric":
        label[i] = "Urban fabric"
      if label[i] =="Discontinuous urban fabric":
        label[i] = "Urban fabric"
      if label[i] == "Non-irrigated arable land":
        label[i] = "Arable land"
      if label[i] == "Permanently irrigated land":
        label[i] = "Arable land"
      if label[i] == "Rice fields":
        label[i] = "Arable land"
      if label[i] == "Vineyards":
        label[i] = "Permanent crops"
      if label[i] == "Fruit trees and berry plantations":
        label[i] = "Permanent crops"
      if label[i] == "Olive groves":
        label[i] = "Permanent crops"
      if label[i] == "Annual crops associated with permanent crops":
        label[i] = "Permanent crops"
      if label[i] == "Natural grassland":
        label[i] = "Natural grassland and sparsely vegetated areas"
      if label[i] == "Sparsely vegetated areas":
        label[i] = "Natural grassland and sparsely vegetated areas"
      if label[i] == "Moors and heathland":
        label[i] = "Moors, heathland and sclerophyllous vegetation"
      if label[i] == "Sclerophyllous vegetation":
        label[i] = "Moors, heathland and sclerophyllous vegetation"
      if label[i] == "Inland marshes":
        label[i] = "Inland wetlands"
      if label[i] == "Peatbogs":
        label[i] = "Inland wetlands"
      if label[i] == "Salt marshes":
        label[i] = "Coastal wetlands"
      if label[i] == "Salines":
        label[i] = "Coastal wetlands"
      if label[i] == "Water courses":
        label[i] = "Inland waters"
      if label[i] == "Water bodies":
        label[i] = "Inland waters"
      if label[i] == "Coastal lagoons":
        label[i] = "Marine waters"
      if label[i] == "Estuaries":
        label[i] = "Marine waters"
      if label[i] == "Sea and ocean":
        label[i] = "Marine waters"
      if label[i] == "Transitional woodland/shrub":
        label[i] = "Transitional woodland, shrub"

    if "Road and rail networks and associated land" in label:
      label.remove("Road and rail networks and associated land")
    if "Sport and leisure facilities" in label:
      label.remove("Sport and leisure facilities")
    if "Port areas" in label:
      label.remove("Port areas")
    if "Airports" in label:
      label.remove("Airports")
    if "Mineral extraction sites" in label:
      label.remove("Mineral extraction sites")
    if "Dump sites" in label:
      label.remove("Dump sites")
    if "Construction sites" in label:
      label.remove("Construction sites")
    if"Green urban areas" in label:
      label.remove("Green urban areas")
    if"Bare rock" in label:
      label.remove("Bare rock")
    if"Burnt areas" in label:
      label.remove("Burnt areas")
    if"Intertidal flats" in label:
      label.remove("Intertidal flats")

    if len(set(label)) < len(label):
      label = set(label)
    if len(label) == 0:
      print(label)

    return sample

In [ ]:
ds["train"] = ds["train"].map(map_labels) # Convert Traning Set to 19-class version

Map:   0%|          | 0/269695 [00:00<?, ? examples/s]

In [ ]:
ds["val"] = ds["val"].map(map_labels) # Convert Validation Set to 19-class version

Map:   0%|          | 0/123723 [00:00<?, ? examples/s]

In [ ]:
ds['test'] = ds['test'].map(map_labels) # Convert Test set to 19-class version

Map:   0%|          | 0/125866 [00:00<?, ? examples/s]

In [ ]:
# This method hot-one encodes labels
encoder_dict = {
        "Urban fabric": 0,
        "Industrial or commercial units": 1,
        "Arable land": 2,
        "Permanent crops": 3,
        "Pastures": 4,
        "Complex cultivation patterns": 5,
        "Land principally occupied by agriculture, with significant areas of natural vegetation": 6,
        "Agro-forestry areas": 7,
        "Broad-leaved forest": 8,
        "Coniferous forest": 9,
        "Mixed forest": 10,
        "Natural grassland and sparsely vegetated areas": 11,
        "Moors, heathland and sclerophyllous vegetation": 12,
        "Transitional woodland, shrub": 13,
        "Beaches, dunes, sands": 14,
        "Inland wetlands": 15,
        "Coastal wetlands": 16,
        "Inland waters": 17,
        "Marine waters": 18
}


def encode(sample):
  new_label = [0] * 19
  label = sample["labels"]
  for i in label:
    new_label[encoder_dict[i]] = 1
  sample["labels"] = new_label
  return sample

In [ ]:
ds["train"] = ds["train"].map(encode) # Hot-one encode training set

Map:   0%|          | 0/269695 [00:00<?, ? examples/s]

In [ ]:
ds["val"] = ds["val"].map(encode) # Hot-one encode validation set

Map:   0%|          | 0/123723 [00:00<?, ? examples/s]

In [ ]:
ds["test"] = ds["test"].map(encode) # Hot-one encode test set

Map:   0%|          | 0/125866 [00:00<?, ? examples/s]

In [ ]:

from keras import layers
import keras

import matplotlib.pyplot as plt
import numpy as np

from tensorflow.keras import layers
import tensorflow.keras.backend as ops
import tensorflow

import matplotlib.pyplot as plt
import numpy as np


#CCT Hyerparameters

#Positional Embedding
positional_emb = True
#Two convultional Layers
conv_layers = 2
#Projection dimension
projection_dim = 128
#Number of Transformer heads
num_heads = 2

transformer_units = [
    projection_dim,
    projection_dim,
]
#Number of Transformer layers
transformer_layers = 2
#Stochastic Depth Rate
stochastic_depth_rate = 0.1
#Learning Rate
learning_rate = 0.001
#Weight Decay
weight_decay = 0.0001
#Batch size
batch_size = 256
#number
num_epochs = 30
#For Random Crop
image_size = 32
#Number of Classes
num_classes = 19
#Image shape
input_shape = (120, 120, 3)

In [ ]:
#All CCT methods were imported from Keras
#https://keras.io/examples/vision/cct/

In [ ]:
class CCTTokenizer(layers.Layer):
    def __init__(
        self,
        kernel_size=3,
        stride=1,
        padding=1,
        pooling_kernel_size=3,
        pooling_stride=2,
        num_conv_layers=conv_layers,
        num_output_channels=[64, 128],
        positional_emb=positional_emb,
        **kwargs,
    ):
        super().__init__(**kwargs)

        # This is our tokenizer.
        self.conv_model = keras.Sequential()
        for i in range(num_conv_layers):
            self.conv_model.add(
                layers.Conv2D(
                    num_output_channels[i],
                    kernel_size,
                    stride,
                    padding="valid",
                    use_bias=False,
                    activation="relu",
                    kernel_initializer="he_normal",
                )
            )
            self.conv_model.add(layers.ZeroPadding2D(padding))
            self.conv_model.add(
                layers.MaxPooling2D(pooling_kernel_size, pooling_stride, "same")
            )

        self.positional_emb = positional_emb

    def call(self, images):
        outputs = self.conv_model(images)
        # After passing the images through our mini-network the spatial dimensions
        # are flattened to form sequences.
        reshaped = keras.ops.reshape(
            outputs,
            (
                -1,
                keras.ops.shape(outputs)[1] * keras.ops.shape(outputs)[2],
                keras.ops.shape(outputs)[-1],
            ),
        )
        return reshaped

In [ ]:
class PositionEmbedding(keras.layers.Layer):
    def __init__(
        self,
        sequence_length,
        initializer="glorot_uniform",
        **kwargs,
    ):
        super().__init__(**kwargs)
        if sequence_length is None:
            self.sequence_length = None
        else:
          self.sequence_length = int(sequence_length)
        self.initializer = keras.initializers.get(initializer)

    def get_config(self):
        config = super().get_config()
        config.update(
            {
                "sequence_length": self.sequence_length,
                "initializer": keras.initializers.serialize(self.initializer),
            }
        )
        return config

    def build(self, input_shape):
        feature_size = input_shape[-1]
        if self.sequence_length is None:
            self.sequence_length = input_shape[-2]
        self.position_embeddings = self.add_weight(
            name="embeddings",
            shape=[self.sequence_length, feature_size],
            initializer=self.initializer,
            trainable=True,
        )

        super().build(input_shape)

    def call(self, inputs, start_index=0):
        shape = keras.ops.shape(inputs)
        feature_length = shape[-1]
        sequence_length = shape[-2]
        # trim to match the length of the input sequence, which might be less
        # than the sequence_length of the layer.
        position_embeddings = keras.ops.convert_to_tensor(self.position_embeddings)
        position_embeddings = keras.ops.slice(
            position_embeddings,
            (start_index, 0),
            (sequence_length, feature_length),
        )
        return keras.ops.broadcast_to(position_embeddings, shape)

    def compute_output_shape(self, input_shape):
        return input_shape

In [ ]:
  class SequencePooling(layers.Layer):
    def __init__(self):
        super().__init__()
        self.attention = layers.Dense(1)

    def call(self, x):
        attention_weights = keras.ops.softmax(self.attention(x), axis=1)
        attention_weights = keras.ops.transpose(attention_weights, axes=(0, 2, 1))
        weighted_representation = keras.ops.matmul(attention_weights, x)
        return keras.ops.squeeze(weighted_representation, -2)

In [ ]:
# Referred from: github.com:rwightman/pytorch-image-models.
class StochasticDepth(layers.Layer):
    def __init__(self, drop_prop, **kwargs):
        super().__init__(**kwargs)
        self.drop_prob = drop_prop
        self.seed_generator = keras.random.SeedGenerator(1337)

    def call(self, x, training=None):
        if training:
            keep_prob = 1 - self.drop_prob
            shape = (keras.ops.shape(x)[0],) + (1,) * (len(x.shape) - 1)
            random_tensor = keep_prob + keras.random.uniform(
                shape, 0, 1, seed=self.seed_generator
            )
            random_tensor = keras.ops.floor(random_tensor)
            return (x / keep_prob) * random_tensor
        return x

In [ ]:
def mlp(x, hidden_units, dropout_rate):
    for units in hidden_units:
        x = layers.Dense(units, activation=keras.ops.gelu)(x)
        x = layers.Dropout(dropout_rate)(x)
    return x

In [ ]:
#Note the rescaling layer. These layers have pre-defined inference behavior.
data_augmentation = keras.Sequential(
    [
        layers.Rescaling(scale=1.0 / 255),
        layers.RandomFlip("horizontal"),
    ],
    name="data_augmentation",
)

In [ ]:
def create_cct_model(
    image_size=image_size,
    input_shape=input_shape,
    num_heads=num_heads,
    projection_dim=projection_dim,
    transformer_units=transformer_units,
):
    inputs = layers.Input(input_shape)

    # Augment data.
    augmented = data_augmentation(inputs)

    # Encode patches.
    cct_tokenizer = CCTTokenizer()
    encoded_patches = cct_tokenizer(augmented)

    # Apply positional embedding.
    if positional_emb:
        sequence_length = encoded_patches.shape[1]
        encoded_patches += PositionEmbedding(sequence_length=sequence_length)(
            encoded_patches
        )

    # Calculate Stochastic Depth probabilities.
    dpr = [x for x in np.linspace(0, stochastic_depth_rate, transformer_layers)]

    # Create multiple layers of the Transformer block.
    for i in range(transformer_layers):
        # Layer normalization 1.
        x1 = layers.LayerNormalization(epsilon=1e-5)(encoded_patches)

        # Create a multi-head attention layer.
        attention_output = layers.MultiHeadAttention(
            num_heads=num_heads, key_dim=projection_dim, dropout=0.1
        )(x1, x1)

        # Skip connection 1.
        attention_output = StochasticDepth(dpr[i])(attention_output)
        x2 = layers.Add()([attention_output, encoded_patches])

        # Layer normalization 2.
        x3 = layers.LayerNormalization(epsilon=1e-5)(x2)

        # MLP.
        x3 = mlp(x3, hidden_units=transformer_units, dropout_rate=0.1)

        # Skip connection 2.
        x3 = StochasticDepth(dpr[i])(x3)
        encoded_patches = layers.Add()([x3, x2])

    # Apply sequence pooling.
    representation = layers.LayerNormalization(epsilon=1e-5)(encoded_patches)
    weighted_representation = SequencePooling()(representation)

    # Classify outputs.
    logits = layers.Dense(num_classes)(weighted_representation)
    # Create the Keras model.
    model = keras.Model(inputs=inputs, outputs=logits)
    return model

In [ ]:
#Create CCT model
model = create_cct_model()

#Create Adam Optimizer with learning Rate and Weight Decay
optimizer = keras.optimizers.AdamW(learning_rate=0.001, weight_decay=0.0001)

#Compile
model.compile(
    optimizer=optimizer,
    loss=keras.losses.BinaryCrossentropy(
        from_logits=True, label_smoothing=0.1
    ),
    metrics=[
        keras.metrics.CategoricalAccuracy(name="accuracy"),
        keras.metrics.TopKCategoricalAccuracy(5, name="top-5-accuracy"),
    ],
)


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
checkpoint_filepath = "'/content/drive/MyDrive/BigEarthFinal/checkpoint.weights.h5" #Create Checkpoint

In [ ]:
#Create Checkpoint call back
checkpoint_callback = keras.callbacks.ModelCheckpoint(
    checkpoint_filepath,
    monitor="val_accuracy",
    save_best_only=True,
    save_weights_only=True,
)

In [ ]:
tf_ds = ds["train"].to_tf_dataset( #Create image generator for Training Set
    columns=["img"],
    label_cols=["labels"],
    shuffle=True,
    batch_size=256,
    num_workers = 4 #Note that that the max number of workers will vary depending on GPU/TPU. 48 workers will work using google collab's GPU
)

/usr/local/lib/python3.11/dist-packages/datasets/arrow_dataset.py:403: FutureWarning: The output of `to_tf_dataset` will change when a passing single element list for `labels` or `columns` in the next datasets version. To return a tuple structure rather than dict, pass a single string.
Old behaviour: columns=['a'], labels=['labels'] -> (tf.Tensor, tf.Tensor)  
             : columns='a', labels='labels' -> (tf.Tensor, tf.Tensor)  
New behaviour: columns=['a'],labels=['labels'] -> ({'a': tf.Tensor}, {'labels': tf.Tensor})  
             : columns='a', labels='labels' -> (tf.Tensor, tf.Tensor) 
  warnings.warn(


In [ ]:
tf_ds_val = ds["val"].to_tf_dataset( #Create image generator for Validation Set
    columns=["img"],
    label_cols=["labels"],
    shuffle=True,
    batch_size=256,
    num_workers = 4 #Note that that the max number of workers will vary depending on GPU/TPU. 48 workers will work using google collab's GPU
)

In [ ]:
# Fit model with 30 epochs and a batch size of 256
history = model.fit(
    tf_ds,
    batch_size=batch_size,
    epochs=30,
    callbacks=[checkpoint_callback],
    validation_data = tf_ds_val
)


Epoch 1/30
1054/1054 ━━━━━━━━━━━━━━━━━━━━ 291s 233ms/step - accuracy: 0.4462 - loss: 0.3655 - top-5-accuracy: 0.7853 - val_accuracy: 0.4809 - val_loss: 0.3352 - val_top-5-accuracy: 0.8653
Epoch 2/30
1054/1054 ━━━━━━━━━━━━━━━━━━━━ 212s 196ms/step - accuracy: 0.4849 - loss: 0.3340 - top-5-accuracy: 0.8595 - val_accuracy: 0.4956 - val_loss: 0.3253 - val_top-5-accuracy: 0.8726
Epoch 3/30
1054/1054 ━━━━━━━━━━━━━━━━━━━━ 212s 196ms/step - accuracy: 0.4884 - loss: 0.3256 - top-5-accuracy: 0.8739 - val_accuracy: 0.4820 - val_loss: 0.3214 - val_top-5-accuracy: 0.8874
Epoch 4/30
1054/1054 ━━━━━━━━━━━━━━━━━━━━ 212s 197ms/step - accuracy: 0.4902 - loss: 0.3211 - top-5-accuracy: 0.8817 - val_accuracy: 0.4933 - val_loss: 0.3178 - val_top-5-accuracy: 0.9028
Epoch 5/30
1054/1054 ━━━━━━━━━━━━━━━━━━━━ 212s 196ms/step - accuracy: 0.4910 - loss: 0.3179 - top-5-accuracy: 0.8867 - val_accuracy: 0.4849 - val_loss: 0.3150 - val_top-5-accuracy: 0.8991
Epoch 6/30
1054/1054 ━━━━━━━━━━━━━━━━━━━━ 212s 196ms/step - 

In [ ]:
tf_ds_test = ds["test"].to_tf_dataset( #Create image generator for Test Set
    columns=["img"],
    label_cols=["labels"],
    shuffle=True,
    batch_size=256,
    num_workers = 4 #Note that that the max number of workers will vary depending on GPU/TPU. 48 workers will work using google collab's GPU
)

In [ ]:
# Load best model weights found during Training
model.load_weights(checkpoint_filepath)
_ , accuracy, top_5_accuracy = model.evaluate(tf_ds_test)
print(f"Test accuracy: {round(accuracy * 100, 2)}%")
print(f"Test top 5 accuracy: {round(top_5_accuracy * 100, 2)}%")

492/492 ━━━━━━━━━━━━━━━━━━━━ 72s 135ms/step - accuracy: 0.5097 - loss: 0.3037 - top-5-accuracy: 0.9140
Test accuracy: 50.97%
Test top 5 accuracy: 91.4%


In [ ]:
model.evaluate(tf_ds_test) #Evalute Model

492/492 ━━━━━━━━━━━━━━━━━━━━ 62s 115ms/step - accuracy: 0.5097 - loss: 0.3037 - top-5-accuracy: 0.9140


[0.3037202060222626, 0.5097007751464844, 0.9139878749847412]

In [ ]:
y_predictions = model.predict(tf_ds_test) # Get Predicted # labels from test set
y_true = ds["test"]["labels"]

492/492 ━━━━━━━━━━━━━━━━━━━━ 68s 118ms/step


In [ ]:
import tensorflow as tf
y_predictions_probabiliities = tf.nn.sigmoid(y_predictions) # Get Sigmoid Activation function results from predicted logits

In [ ]:
y_predictions_probabiliities

<tf.Tensor: shape=(125866, 19), dtype=float32, numpy=
array([[0.06917268, 0.03504444, 0.8235986 , ..., 0.05881923, 0.11603456,
        0.04388122],
       [0.42132154, 0.04665869, 0.9722571 , ..., 0.04867497, 0.03090381,
        0.04603129],
       [0.03582156, 0.03949663, 0.49553642, ..., 0.04778549, 0.9772857 ,
        0.04491984],
       ...,
       [0.09412542, 0.03766977, 0.78102833, ..., 0.05171735, 0.14912303,
        0.04157107],
       [0.03925664, 0.04035242, 0.0636957 , ..., 0.04859553, 0.51539874,
        0.04024595],
       [0.04967981, 0.05341111, 0.05447578, ..., 0.04426669, 0.05043306,
        0.950639  ]], dtype=float32)>

In [ ]:
y_predictions = np.array(y_predictions_probabiliities) # Convert prediction probabilties to numpy array

In [ ]:
y_predictions

array([[0.06917268, 0.03504444, 0.8235986 , ..., 0.05881923, 0.11603456,
        0.04388122],
       [0.42132154, 0.04665869, 0.9722571 , ..., 0.04867497, 0.03090381,
        0.04603129],
       [0.03582156, 0.03949663, 0.49553642, ..., 0.04778549, 0.9772857 ,
        0.04491984],
       ...,
       [0.09412542, 0.03766977, 0.78102833, ..., 0.05171735, 0.14912303,
        0.04157107],
       [0.03925664, 0.04035242, 0.0636957 , ..., 0.04859553, 0.51539874,
        0.04024595],
       [0.04967981, 0.05341111, 0.05447578, ..., 0.04426669, 0.05043306,
        0.950639  ]], dtype=float32)

In [ ]:
y_predictions_encoded = (y_predictions > 0.5).astype(int) # If probabilites for a label is greater than 0.5, make that label 1

In [ ]:
y_predictions_encoded[3]

array([0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [ ]:
# Calculate Metrics
from sklearn.metrics import precision_score, recall_score, f1_score
# Get micro precision score
print(precision_score(y_true, y_predictions_encoded, average='micro'))
# Get micro recall score
print(recall_score(y_true, y_predictions_encoded, average='micro'))
# Get micro f1 score
print(f1_score(y_true, y_predictions_encoded, average='micro'))


0.26052155951458816
0.23622101462248432
0.24777689823134125


In [ ]:
model.summary() #Model Summary

Model: "functional_2"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 120, 120,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ data_augmentation   │ (None, 120, 120,  │          0 │ input_layer[0][0] │
│ (Sequential)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ cct_tokenizer       │ (None, 900, 128)  │     75,456 │ data_augmentatio… │
│ (CCTTokenizer)      │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ position_embedding  │ (None, 900, 128)  │    115,200 │ cct_tokenizer[0]… │
│ (PositionEmbedding) │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add (Add)           │ (None, 900, 128)  │          0 │ cct_tokenizer[0]… │
│                     │                   │            │ position_embeddi… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ layer_normalization │ (None, 900, 128)  │        256 │ add[0][0]         │
│ (LayerNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ multi_head_attenti… │ (None, 900, 128)  │    131,968 │ layer_normalizat… │
│ (MultiHeadAttentio… │                   │            │ layer_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ stochastic_depth    │ (None, 900, 128)  │          0 │ multi_head_atten… │
│ (StochasticDepth)   │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add_1 (Add)         │ (None, 900, 128)  │          0 │ stochastic_depth… │
│                     │                   │            │ add[0][0]         │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ layer_normalizatio… │ (None, 900, 128)  │        256 │ add_1[0][0]       │
│ (LayerNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (Dense)       │ (None, 900, 128)  │     16,512 │ layer_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 900, 128)  │          0 │ dense[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_1 (Dense)     │ (None, 900, 128)  │     16,512 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 900, 128)  │          0 │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ stochastic_depth_1  │ (None, 900, 128)  │          0 │ dropout_2[0][0]   │
│ (StochasticDepth)   │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add_2 (Add)         │ (None, 900, 128)  │          0 │ stochastic_depth… │
│                     │                   │            │ add_1[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ layer_normalizatio… │ (None, 900, 128)  │        256 │ add_2[0][0]       │
│ (LayerNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ multi_head_attenti… │ (None, 900, 128)  │    131,968 │ layer_normalizat… │
│ (MultiHeadAttentio… │                   │            │ layer_normalizat

 Total params: 1,573,502 (6.00 MB)

 Trainable params: 524,500 (2.00 MB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 1,049,002 (4.00 MB)

# BigEarthNet Classification using CCT

This project demonstrates the classification of satellite images from the BigEarthNet dataset using a Compact Convolutional Transformer (CCT) model implemented with Keras. The BigEarthNet dataset is a large-scale archive of Sentinel-2 satellite images with multi-labels, representing various land cover classes.

## Experiment Overview

The goal of this experiment is to train a CCT model to classify the land cover in the BigEarthNet dataset. The experiment involves:

1.  **Data Loading and Preprocessing**: Loading the BigEarthNet dataset from the Hugging Face Hub, converting the 43-class labels to a 19-class version, and hot-one encoding the labels.
2.  **Model Definition**: Implementing the CCT architecture using custom Keras layers for tokenization, positional embedding, sequence pooling, and stochastic depth.
3.  **Model Training**: Training the CCT model on the preprocessed training data using the AdamW optimizer and Binary Crossentropy loss, with a ModelCheckpoint callback to save the best weights based on validation accuracy.
4.  **Model Evaluation**: Evaluating the trained model on the test set using accuracy, top-5 accuracy, precision, recall, and F1-score.

## Dataset

The BigEarthNet dataset used in this experiment is sourced from the Hugging Face Hub: `danielz01/BigEarthNet-S2-v1.0`. The dataset contains Sentinel-2 satellite images and associated land cover labels. The original 43 classes are mapped to a reduced set of 19 classes for this experiment.

## Model Architecture

The model used is a Compact Convolutional Transformer (CCT). The key components of the CCT implemented here include:

-   **CCT Tokenizer**: A convolutional network to extract features from image patches and flatten them into sequences.
-   **Position Embedding**: Adds positional information to the tokenized sequences.
-   **Transformer Layers**: Multiple layers of multi-head attention and MLP blocks with stochastic depth for regularization.
-   **Sequence Pooling**: A pooling layer to aggregate the sequence representation into a single vector.
-   **Classifier**: A dense layer for multi-label classification.

The model hyperparameters are defined in the notebook and can be adjusted for further experimentation.

## Results

After training for 30 epochs, the model achieved the following metrics on the test set:

-   **Test Accuracy**: 50.97%
-   **Test Top 5 Accuracy**: 91.40%
-   **Micro Precision**: 0.261
-   **Micro Recall**: 0.236
-   **Micro F1 Score**: 0.248

These results indicate that the model has some capability in classifying the land cover, but there is room for improvement, especially in terms of precision, recall, and F1 score, which are more sensitive to multi-label classification performance than accuracy.

## Setup and Usage

To run this experiment, you will need:

-   Google Colab environment (recommended for GPU access)
-   Python 3.7+
-   Libraries: `tensorflow`, `keras`, `datasets`, `scikit-learn`, `matplotlib`, `numpy`

The necessary libraries are installed in the notebook using `pip`.

You will also need to log in to the Hugging Face Hub to access the dataset. This can be done using `huggingface-cli login` and providing your Hugging Face token.

The notebook contains all the code required to:

1.  Install dependencies.
2.  Load and preprocess the dataset.
3.  Define and compile the CCT model.
4.  Train the model with a checkpoint callback.
5.  Evaluate the model and print the results.

## Future Work

Potential areas for improvement and future work include:

-   Experimenting with different CCT hyperparameters.
-   Exploring other data augmentation techniques.
-   Investigating different loss functions and optimizers.
-   Evaluating the model on the original 43-class problem.
-   Comparing the CCT model with other deep learning architectures for satellite image classification.
-   Analyzing the misclassifications to identify challenging land cover classes.

## License

This project is licensed under the [Specify License, e.g., MIT License].

## Acknowledgements

-   The BigEarthNet dataset creators.
-   The Keras team for the CCT implementation example.

# BigEarthNet Classification using CCT

This project demonstrates the classification of satellite images from the BigEarthNet dataset using a Compact Convolutional Transformer (CCT) model implemented with Keras. The BigEarthNet dataset is a large-scale archive of Sentinel-2 satellite images with multi-labels, representing various land cover classes.

## Experiment Overview

The goal of this experiment is to train a CCT model to classify the land cover in the BigEarthNet dataset. The experiment involves:

1.  **Data Loading and Preprocessing**: Loading the BigEarthNet dataset from the Hugging Face Hub, converting the 43-class labels to a 19-class version, and hot-one encoding the labels.
2.  **Model Definition**: Implementing the CCT architecture using custom Keras layers for tokenization, positional embedding, sequence pooling, and stochastic depth.
3.  **Model Training**: Training the CCT model on the preprocessed training data using the AdamW optimizer and Binary Crossentropy loss, with a ModelCheckpoint callback to save the best weights based on validation accuracy.
4.  **Model Evaluation**: Evaluating the trained model on the test set using accuracy, top-5 accuracy, precision, recall, and F1-score.

## Dataset

The BigEarthNet dataset used in this experiment is sourced from the Hugging Face Hub: `danielz01/BigEarthNet-S2-v1.0`. The dataset contains Sentinel-2 satellite images and associated land cover labels. The original 43 classes are mapped to a reduced set of 19 classes for this experiment using the `map_labels` function. The labels are then hot-one encoded using the `encode` function.

## Model Architecture

The model used is a Compact Convolutional Transformer (CCT). The key components of the CCT implemented here include:

-   **CCT Tokenizer**: A convolutional network to extract features from image patches and flatten them into sequences.
-   **Position Embedding**: Adds positional information to the tokenized sequences.
-   **Transformer Layers**: Multiple layers of multi-head attention and MLP blocks with stochastic depth for regularization.
-   **Sequence Pooling**: A pooling layer to aggregate the sequence representation into a single vector.
-   **Classifier**: A dense layer for multi-label classification.

The model hyperparameters are defined in the notebook and can be adjusted for further experimentation.

## Results

After training for 30 epochs, the model achieved the following metrics on the test set:

-   **Test Accuracy**: 50.97%
-   **Test Top 5 Accuracy**: 91.40%
-   **Micro Precision**: 0.261
-   **Micro Recall**: 0.236
-   **Micro F1 Score**: 0.248

These results indicate that the model has some capability in classifying the land cover, but there is room for improvement, especially in terms of precision, recall, and F1 score, which are more sensitive to multi-label classification performance than accuracy.

## Setup and Usage

To run this experiment, you will need:

-   Google Colab environment (recommended for GPU access)
-   Python 3.7+
-   Libraries: `tensorflow`, `keras`, `datasets`, `scikit-learn`, `matplotlib`, `numpy`

The necessary libraries are installed in the notebook using `pip`.

You will also need to log in to the Hugging Face Hub to access the dataset. This can be done using `huggingface-cli login` and providing your Hugging Face token.

The notebook contains all the code required to:

1.  Install dependencies.
2.  Load and preprocess the dataset.
3.  Define and compile the CCT model.
4.  Train the model with a checkpoint callback.
5.  Evaluate the model and print the results.

## Future Work

Potential areas for improvement and future work include:

-   Experimenting with different CCT hyperparameters.
-   Exploring other data augmentation techniques.
-   Investigating different loss functions and optimizers.
-   Evaluating the model on the original 43-class problem.
-   Comparing the CCT model with other deep learning architectures for satellite image classification.
-   Analyzing the misclassifications to identify challenging land cover classes.

## License

This project is licensed under the [Specify License, e.g., MIT License].

## Acknowledgements

-   The BigEarthNet dataset creators.
-   The Keras team for the CCT implementation example.